File: languages/typescript/src/smells/rules/no-data-clumps.ts
import { defineRule } from "@oxlint/plugins";
import type { ESTree } from "@oxlint/plugins";

import { numberOption, thresholdSchema } from "../shared/options.ts";

const DEFAULT_MIN_GROUP = 3;

type ParameterOwner = ESTree.ArrowFunctionExpression | ESTree.Function;

function parameterNames(node: ParameterOwner): string[] {
	return node.params
		.map((parameter) => (parameter.type === "Identifier" ? parameter.name : null))
		.filter((name): name is string => name !== null);
}

/** Ban the same group of parameters travelling between functions. */
export const noDataClumpsRule = defineRule({
	meta: {
		type: "problem",
		docs: {
			description:
				"Disallow the same group of `minGroup` parameter names appearing in more than one function; values that travel together belong to one type.",
		},
		messages: {
			dataClump:
				"This function and an earlier one both take {{names}}. Values that travel together belong to one thing; give them a type and pass that instead.",
		},
		...thresholdSchema({ minGroup: DEFAULT_MIN_GROUP }),
	},
	createOnce(context) {
		const seen = new Map<string, true>();

		const check = (node: ParameterOwner) => {
			const minimum = numberOption(context.options, "minGroup", DEFAULT_MIN_GROUP);
			const names = parameterNames(node);
			if (names.length < minimum) return;
			const key = [...names].sort().join(",");
			if (!seen.has(key)) {
				seen.set(key, true);
				return;
			}
			context.report({
				node,
				messageId: "dataClump",
				data: { names: [...names].sort().map((name) => `\`${name}\``).join(", ") },
			});
		};

		return {
			Program() {
				seen.clear();
			},
			ArrowFunctionExpression: check,
			FunctionDeclaration: check,
			FunctionExpression: check,
		};
	},
});

File: skills/install-slop-guard/assets/python/rules/smells/no_data_clumps.py
"""Reject the same group of parameters travelling between functions."""

from __future__ import annotations

import ast

from ...rule import FunctionNode, Rule

DEFAULT_MIN_GROUP = 3
RECEIVERS = frozenset({"self", "cls", "mcs"})


class NoDataClumps(Rule):
    name = "no-data-clumps"
    group = "smells"
    reference = "https://refactoring.guru/smells/data-clumps"
    description = (
        "Disallow the same group of `min_group` parameter names, 3 by default, appearing in more "
        "than one function in a module."
    )

    def run(self) -> list:
        minimum = self.threshold("min_group", DEFAULT_MIN_GROUP)
        seen: dict[frozenset[str], tuple[str, ast.AST]] = {}
        for node in ast.walk(self.source.tree):
            if not isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef)):
                continue
            names = frozenset(
                parameter.arg for parameter in self.parameters(node) if parameter.arg not in RECEIVERS
            )
            if len(names) < minimum:
                continue
            earlier = seen.get(names)
            if earlier is None:
                seen[names] = (node.name, node)
                continue
            listed = ", ".join(f"`{name}`" for name in sorted(names))
            self.report(
                node,
                f"`{node.name}` and `{earlier[0]}` both take {listed}. Values that travel "
                f"together belong to one thing; give them a type and pass that instead.",
            )
        return self.diagnostics
